### DISCUSSION TITLE
Encryption?

### DISCUSSION THREAD

**Discussion by noahbliss** [ExpressLRS/ExpressLRS#1609] [Q&A]
It seems there wasn't much information given in regards to protection of data over the air. In event of a collision of the bind phrase it would be easy to hijack or at least DoS a (potentially very expensive) drone. Has any thought been given to encryption of data over the air?

**Comment by JyeSmith** [comment]
Theres not much you could do even if it was encrypted.  Anyone could spam the spectrum with interference and FS (DoS) a link.

Some further reading here https://github.com/ExpressLRS/ExpressLRS/pull/1411

**Reply by EhAye** [reply]
i think the difference is jamming is illegal, and it simply does a GPS RTH 
vs someone hijacking the link and flying it to a spot where they can pick it up and drive off with it.

i suppose we could use really strange/nonstandard channel mappings, so manual flight is next to impossible without the correct radio setup

**Reply by noahbliss** [reply]
The solution for manual radio hijacking must be encryption. Anything else is just kicking the can. Since ELRS is the leading open standard, it will be the primarily attacked protocol. Given the code is open source, reverse engineering is a feature, not a bug. Channel hopping would be useful for detecting and remediating jamming (in most common cases, from other elrs users inadvertently) but even proprietary channel hopping as a security mechanism has a history of being easy to break. 

A TBS hack is detailed here:
https://hackaday.com/2022/02/20/reverse-engineering-a-900-mhz-rc-transmitter-and-receiver/

**Comment by noahbliss** [comment]
When we're talking about MAVLINK over ELRS, the abuse potential gets much much more sinister. With the war in Ukraine, Russians are using signal intelligence to sniff off-the-shelf drone traffic (e.g. DJI) and extract the RTL coordinates from them. Mortar rounds kill the pilot minutes later. Security cannot be an afterthought in any modern technology.

**Reply by JyeSmith** [reply]
Thats terrible to hear.

ELRS does not use mavlink, but more importantly ELRS is not aimed at or looking to support military applications.

**Reply by smscat** [reply]
We are (ukrainians) looking to use any technology to protect our country and our people from russian aggression. We have to use anything we have, including elrs. And we will. With or without community help.

**Comment by brandonrc** [comment]
Encryption is also illegal on radio control links in a few countries. 

https://fieldradio.org/can-ham-radio-be-encrypted/

I recommend the rfd900x if you want a nice mavlink/encryption radio. They are very solid.

Having encryption would take away rhe express out of express lrs

**Reply by commanderguy3001** [reply]
actually, in this case, I bet it would be allowed, as expresslrs is a lot closer to wifi communication, which is totally allowed to be encrypted in any way you like, as it isn't ham communication, and the whole system is on an ISM band too (e.g. people talking to each other)

**Reply by noahbliss** [reply]
I was more commenting due to the growing scope of ELRS in including telemetry. The firmware for rfd900x is unfortunately closed-source, so a good open source radio with long range that supports encryption seems to be an unfilled market corner at the moment. Thanks for the feedback though.

**Reply by sensei-hacker** [reply]
> Having encryption would take away rhe express out of express lrs

Not at all if done anything like correctly.
The slow part that takes several milliseconds happens when you bind / power up.

If devs don't want encryption for *some other reason*, that's cool. Two or three CPU cycles doing XOR is not a reason.

**Reply by sensei-hacker** [reply]
> Encryption is also illegal on radio control links in a few countries.
> 
> https://fieldradio.org/can-ham-radio-be-encrypted/
> 
> I recommend the rfd900x if you want a nice mavlink/encryption radio. They are very solid.
> 
> Having encryption would take away rhe express out of express lrs


ELRS isn't ham (licensed user)..ELRS is ISM. It's not useful to talk about regulations for licensed hams like myself, operating on our ham bands, when the topic is ISM band usage by unlicensed persons.

The regulations that apply to the ISM bands are - the ISM regulations. Possibly the most popular use of ISM now is LORA. You may discover that LORAWAN is in fact encrypted.

Basically you can do whatever you want on ISM except send voice or transmit longer than allowed at a given output power and dwell.

**Reply by gluedog** [reply]
What are you talking about, "encryption is illegal" ?
They've found a vulnerability that can be easily patched over a year ago and yet you guys are doing nothing, instead of posting stupid links about how "encryption is illegal" (imagine that) why don't you encrypt the protocol already

**Reply by brandonrc** [reply]
1. I was wrong about the legality of encrypting control links for drones. Lots of the RF we do in this hobby falls under the HAM regulations. That is why you have to get your HAM license to operate in some conditions.  But again I was wrong. 
2. We do not want to encrypt our control link because there is no benefit for the intended use cases. 
3. You can stop complaining on github and make your own encrypted radio link if you want it so bad... or how about this, make a pull request that shows something of value before complaining about how we spend our free time helping this community.

**Comment by sensei-hacker** [comment]
@smscat  Please contact me privately at deepmagicbeginshere@gmail.com

**Reply by MykhailoVyn** [reply]
Hi @sensei-hacker,
May I also contact you privately?

**Comment by sensei-hacker** [comment]
> Hi @sensei-hacker, May I also contact you privately?

Absolutely!
I have a significant amount of work done on this.  I had to set it down for a week to while I did another project, but it's in progress.

**Reply by gittb** [reply]
Do you plan to release this publicly?

**Comment by sensei-hacker** [comment]
> Do you plan to release this publicly?

Yes. I'm working on it a bit more today, then it should be ready for testing. It's working better than I had expected. There's virtually no performance impact.

That's partly because I coded it very carefully, partly because the processor runs at hundreds of millions of cycles per second, while the radio only sends a few hundred packets per second.  So the processor ends up sitting around waiting for the radio anyway. :)

**Reply by gluedog** [reply]
The newer flight controllers run at 30% usage most of the time anyway. I'm surprised that the vulnerability that was found over a year ago hasn't been patched yet

**Reply by gluedog** [reply]
What are the devs doing anyway?

**Comment by JyeSmith** [comment]
Closing.

This is not a feature the ELRS project is looking to support at this time.

**Reply by hasselk** [reply]
If Encryption is not a topic at the moment, REAL Key Based AUTHENTICATION Should realy be one

**Reply by commanderguy3001** [reply]
@hasselk if you need it so badly, there's no one stopping you from adding it and making a PR. no one needed it until now, which is one of the big reasons that it isn't a thing. there's also some technical limitations (having no single free bit left in the rf protocol, and increasing the packet size would decrease sensitivity and therefor range). but I bet you if you make it happen anyway, there will be a lot of people who would not complain about it.
